pandas/core/indexes/datetimelike.py

```diff
@@ -728,31 +570,83 @@ def _maybe_box_as_values(self, values, **attribs):
+
+        return self
+
     def astype(self, dtype, copy=True):
-        if is_object_dtype(dtype):
-            return self._box_values_as_index()
-        elif is_string_dtype(dtype) and not is_categorical_dtype(dtype):
-            return Index(self.format(), name=self.name, dtype=object)
-        elif is_integer_dtype(dtype):
-            # TODO(DatetimeArray): use self._values here.
-            # Can't use ._values currently, because that returns a
-            # DatetimeIndex, which throws us in an infinite loop.
-            return Index(self.values.astype('i8', copy=copy), name=self.name,
-                         dtype='i8')
-        elif (is_datetime_or_timedelta_dtype(dtype) and
-              not is_dtype_equal(self.dtype, dtype)) or is_float_dtype(dtype):
-            # disallow conversion between datetime/timedelta,
-            # and conversions for any datetimelike to float
-            msg = 'Cannot cast {name} to dtype {dtype}'
-            raise TypeError(msg.format(name=type(self).__name__, dtype=dtype))
-        return super(DatetimeIndexOpsMixin, self).astype(dtype, copy=copy)
+        # NB: moved from PeriodIndex
+        new_values = self._values.astype(dtype, copy=copy)
+        return Index(new_values, dtype=dtype, name=self.name)
+
+    def view(self, dtype=None, type=None):
+        # NB: moved from PeriodIndex
+        if dtype is None or dtype is __builtins__['type'](self):
+            # Series.copy() eventually calls this. Need to call
+            # _shallow_copy here so that we don't propagate modifications
+            # to attributes like .index.name
+            return self._shallow_copy()
+        return self._ndarray_values.view(dtype=dtype)
+
+    @deprecate_kwarg(old_arg_name='n', new_arg_name='periods')
+    def shift(self, periods, freq=None):
+        # NB: moved from PeriodIndex
+        """
+        Shift index by desired number of increments.
+
+        This method is for shifting the values of period indexes
+        by a specified time increment.
+
+        Parameters
+        ----------
+        periods : int, default 1
+            Number of periods (or increments) to shift by,
+            can be positive or negative.
+
+            .. versionchanged:: 0.24.0
+        freq :
+
+        Returns
+        -------
+        pandas.PeriodIndex
+            Shifted index.
+
+        See Also
+        --------
+        DatetimeIndex.shift : Shift values of DatetimeIndex.
+        """
+        new_values = self._data._time_shift(periods, freq=freq)
+        return self._simple_new(new_values, name=self.name, freq=self.freq)
 
     @Appender(DatetimeLikeArrayMixin._time_shift.__doc__)
     def _time_shift(self, periods, freq=None):
         result = DatetimeLikeArrayMixin._time_shift(self, periods, freq=freq)
         result.name = self.name
         return result
 
+    # -
+    # dispatch
+
+    def _has_same_tz(self, other):
+        return self._data._has_same_tz(other)
```

Comment:
should be in DatetimeIndex?  (and probably inherited via the accessor that I'm guessing will appear when I scroll down some more?)